sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -308,54 +300,93 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+        Engine.IndexResult indexResult = index(seqNo, primaryTerm, version, isRetry, indexShard, item);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            Map<String, Object> source = JsonXContent.jsonXContent.createParser(
+                NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                BytesReference.toBytes(rawSource)).map();
+            returnvalues = returnGen.generateReturnValues(
+                new Doc(-1,
+                        indexShard.shardId().getIndexName(),
+                        item.id(),
+                        indexResult.getVersion(),
+                        indexResult.getSeqNo(),
+                        primaryTerm,
+                        source,
+                    () -> ""
+                )
             );
+        }
+        return new IndexItemResponse(indexResult.getTranslogLocation(), returnvalues);
+    }
 
-            if (item.returnValues() != null) {
-                updatedDoc = currentDoc.withUpdatedSource(updatedSource);
-            }
-
-            item.source(BytesReference.bytes(XContentFactory.jsonBuilder().map(updatedSource)));
-            seqNo = item.seqNo();
-            primaryTerm = item.primaryTerm();
-            version = Versions.MATCH_ANY;
+    protected IndexItemResponse update(@Nullable ReturnValueGen returnGen,
+                                       @Nullable UpdateSourceGen updateSourceGen,
+                                       IndexShard indexShard,
+                                       ShardUpsertRequest.Item item,
+                                       boolean isRetry) throws Exception {
+        assert updateSourceGen != null : "UpdateSourceGen must not be null";
+        Doc fetchedDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
+        Map<String, Object> source = updateSourceGen.generateSource(
+            fetchedDoc,
+            item.updateAssignments(),
+            item.insertValues()
+        );
+        BytesReference rawSource = BytesReference.bytes(XContentFactory.jsonBuilder().map(source));
+        item.source(rawSource);
+        long seqNo = item.seqNo();
+        long primaryTerm = item.primaryTerm();
+        long version = Versions.MATCH_ANY;
+
+        Engine.IndexResult indexResult = index(seqNo, primaryTerm, version, isRetry, indexShard, item);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            returnvalues = returnGen.generateReturnValues(
+                new Doc(fetchedDoc.docId(),
+                        fetchedDoc.getIndex(),
+                        fetchedDoc.getId(),
+                        indexResult.getVersion(),
+                        indexResult.getSeqNo(),
+                        primaryTerm,
+                        source,
+                    () -> "")
```

Comment:
I think we should adapt checkstyle rules for lambda, this is what checkstyle is asking for, but it seems odd to me.